internal/metrics/metrics.go

```diff
@@ -0,0 +1,102 @@
+package metrics
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/push"
+	"github.com/sirupsen/logrus"
+)
+
+var ErrorMetricsAlreadyRunning = errors.New("metrics server is already running")
+
+// Pusher is a small wrapper around prometheus's Pusher to allow you to easily set up metrics to be pushed to a prometheus-pushgateway. It is thread safe
+type Pusher struct {
+	pusher *push.Pusher
+	mu     sync.Mutex
+	wg     sync.WaitGroup
+	live   bool
+	log    *logrus.Logger
+}
+
+func NewPusher(logger *logrus.Logger, task string) *Pusher {
+	newPusher := &Pusher{
+		mu:     sync.Mutex{},
+		wg:     sync.WaitGroup{},
+		live:   false,
+		log:    logger,
+		pusher: push.New("http://fleet-pushgateway:9090/", task),
+	}
+
+	return newPusher
+}
+
+// Add a new collector to the pusher
+func (mp *Pusher) AddCollector(collector prometheus.Collector) {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	mp.pusher = mp.pusher.Collector(collector)
+}
+
+// Start the metrics pusher in a background thread
+func (mp *Pusher) Start() error {
+	if mp.IsAlive() {
+		return ErrorMetricsAlreadyRunning
+	}
+
+	// First test that the pushgateway is accessible by running an initial push
+	err := mp.Push()
+	if err != nil {
+		return err
+	}
+
+	mp.wg.Add(1)
+
+	go func() {
+		defer mp.wg.Done()
+
+		mp.mu.Lock()
+		mp.live = true
+		mp.mu.Unlock()
+
+		for mp.IsAlive() {
+			time.Sleep(time.Second)
+
+			err := mp.Push()
+			if err != nil {
+				mp.log.Errorf("Failed to push metrics to gateway: %s", err.Error())
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// Used to check if the metrics pusher is still running
+func (mp *Pusher) IsAlive() bool {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	return mp.live
+}
+
+// Kills the metrics pusher thread and waits for it to die
+func (mp *Pusher) KillAndWait() {
+	mp.mu.Lock()
+	mp.live = false
+	mp.mu.Unlock()
+
+	mp.wg.Wait()
+}
+
+// Pushes metrics to the pushgateway
+func (mp *Pusher) Push() error {
+	mp.mu.Lock()
+	defer mp.mu.Unlock()
+
+	return mp.pusher.Push()
```

Comment:
Where do we define what actually gets pushed here?

Reply:
The collector. So i made it a bit "dynamic". Maybe a bit too dynamic. You can see on cmd/inventory.go:53 I add the Go Collector. That is what is being pushed. IT collects all the default go data about memstats, gc, and cpu.

Reply:
What are you your plans for app-specific metrics?

Reply:
Tasks completed, tracking each type of task individually as we add more

For inventory task specifically I think the only thing I could record is condition creation counts.

Reply:
Also, thinking about a more go like way to do the pusher thread. I dont like any of my options. Using the ctx seems the normal go way to do things, problem is then its not self contained. I would have to make changes to how we handle cancel ctx and stuff. I feel that the C way is the way (it always has been).